framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -2653,6 +2653,58 @@ public AnnotatedWildcardType getUninferredWildcardType(AnnotatedTypeVariable typ
         return wctype;
     }
 
+    /**
+     * Returns the most specific base type of {@code annotatedTypeMirror} whose erasure type
+     * is the upper bound of {@code wildcard}, if {@code wildcard}'s upper
+     * bound is a super type of {@code annotatedTypeMirror}.
+     *
+     * Otherwise, returns {@code annotatedTypeMirror} unmodified.
+     *
+     * For example:
+     * <pre>
+     * wildcard := @NonNull ? extends @NonNull Object
+     * annotatedTypeMirror := @Nullable String
+     *
+     * widenToUpperBound(annotatedTypeMirror, wildcard) returns @Nullable String
```

Comment:
Shouldn't this return Nullable Object? If this is the right answer, can you also give an example where the return type is different from annotatedTypeMirror?